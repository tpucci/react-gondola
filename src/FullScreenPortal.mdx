---
name: FullScreenPortal
menu: API
route: /api/full-screen-portal
---

# FullScreenPortal

`type: React.Component`

## Props

| Prop name  | Required | Default     | Usage                                                                                                      |
| :--------: | :------: | ----------- | ---------------------------------------------------------------------------------------------------------- |
| `children` |  `true`  | `undefined` | Children are wrapped in a `React.Fragment`. All full screen Stops are rendered in front of those children. |

## Properties

### `fullScreenStack: IStreamListener<StopInterface[]>`

MobX Strem Listener of list of React Gondola full screen Stops. This stream is observed by MobX and fires the rendering of the full screen stack in the `FullScreenPortal.render` method.

## Basic usage

Wrap your application main canal with `FullScreenPortal`. `FullScreenPortal` must appear at the top of your tree in order to render full screen pages.

```tsx
import React, { Component } from 'react';
import { FullScreenPortal } from 'react-nonav';

import { MainCanal } from './canals/MainCanal';

export default class App extends Component {
  render() {
    return (
      <FullScreenPortal>
        <MainCanal />
      </FullScreenPortal>
    );
  }
}
```
