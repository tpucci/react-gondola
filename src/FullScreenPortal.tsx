import React, { Component as ReactComponent, Fragment } from 'react';
import { View, StyleSheet } from 'react-native';
import { Observer } from 'mobx-react/native';
import { fromStream } from 'mobx-utils';
import { Navigation } from './Navigation';

export class FullScreenPortal extends ReactComponent {
  fullScreenStack = fromStream(Navigation.getInstance().fullSceenStack$);

  render() {
    return (
      <View style={StyleSheet.absoluteFill}>
        <Fragment>{this.props.children}</Fragment>
        <Observer>
          {() => {
            if (this.fullScreenStack.current) {
              return (
                this.fullScreenStack.current &&
                this.fullScreenStack.current.map(({ Component, name }) => (
                  <View style={StyleSheet.absoluteFill} key={name}>
                    <Component />
                  </View>
                ))
              );
            }
            return null;
          }}
        </Observer>
      </View>
    );
  }
}
