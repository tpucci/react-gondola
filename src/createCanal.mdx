---
name: createCanal
menu: API
route: /api/create-canal
---

# createCanal

`type: <Authorizations = { [K in Stop['name']]?: boolean }>(stopsList: Stop[]): ComponentType<CanalComponentProps<Authorizations>>`

`createCanal` takes a list of `Stop` and returns a new canal. A valid `Stop` is an object with a unique `name` key, a component key and an optional `isFullScreen` flag. See `Stop` for more details.

## Stop

|    Key name    | Required | Default     | Usage                                                                           |
| :------------: | :------: | ----------- | ------------------------------------------------------------------------------- |
|     `name`     |  `true`  | `undefined` | Unique identifier.                                                              |
|  `Component`   |  `true`  | `undefined` | The component that will be rendered when the user progress reaches that stop.   |
|    `props`     | `false`  | `undefined` | Optional. Props that will be passed down to the component.                      |
|    `onBack`    | `false`  | `undefined` | Optional. Android Back button Callback.                                         |
| `isFullScreen` | `false`  | `undefined` | Optional. Indicates if the stop needs to be rendered in the `FullScreenPortal`. |

## Canal Props

|       Prop name       | Required | Default                               | Usage                                                                           |
| :-------------------: | :------: | ------------------------------------- | ------------------------------------------------------------------------------- |
|        `style`        | `false`  | `ReactNative.StyleSheet.absoluteFill` | Style applied to the Canal wrapper.                                             |
| `[K in Stop['name']]` | `false`  | `undefined`                           | Indicates to the canal that we are authorized to render the corresponding stop. |

## Canal Properties

### `canalId: string`

The canal unique id generated with `Date.now().toString()`

### `stopsList: Stop[]`

The list of Stop given as the `createCanal` first argument.

### `authorizations: Subject<Authorizations>`

RxJS Subject of Authorizations.

### `progress: Observable<Stop[]>`

RxJS Observable of the list of authorized stops which represent a valid route. A valid route starts from the first stop, go through each stop and ends at the first non-authorized stop. If the first stop is not authorized, the progress emits an empty list.

### `stack: IStreamListener<Stop[]>`

MobX Strem Listener of list of local Stops. All full screen stops are filtered out. This stream is observed by MobX and fires the rendering of the canal stack in the `Canal.render` method.

## Basic usage

Link your pages in a canal.

```tsx
import React from "react";
import { createCanal } from "react-gondola";
import { FirstName } from "./FirstName";
import { LastName } from "./LastName";
import { Confirm } from "./Confirm";

export const SignInCanal = createCanal([
  { name: "firstName", Component: FirstName },
  { name: "lastName", Component: LastName },
  { name: "confirm", Component: Confirm, isFullScreen: true }
]);
```
