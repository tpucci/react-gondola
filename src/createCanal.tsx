import React, { ComponentType, Component as ReactComponent } from 'react';
import { ViewStyle, View, StyleSheet, StyleProp } from 'react-native';
import { Observer } from 'mobx-react/native';
import { fromStream } from 'mobx-utils';
import { Subject, Observable } from 'rxjs';
import {
  map,
  distinctUntilChanged,
  withLatestFrom,
  share
} from 'rxjs/operators';
import { Navigation } from './Navigation';
import { last } from './utils/Array.last';
import { IBackEvent } from './Navigation/BackHandlerDelegate';
import { withBackContext, WithBackContext } from './withBackContext';
import { StopValidator, IStop } from './StopValidator';
import { StopHOC } from './StopHOC';

type CanalComponentProps<T> = {
  style?: StyleProp<ViewStyle>;
} & T;

export interface ICanal {
  back$: Observable<IBackEvent>;
}

export const createCanal = <
  StopName extends string,
  Stop extends IStop<StopName>,
  Authorizations = { [K in Stop['name']]?: boolean }
>(
  stops: Stop[]
): ComponentType<CanalComponentProps<Authorizations>> => {
  StopValidator.validateList(stops);

  class CanalComponent
    extends ReactComponent<WithBackContext<CanalComponentProps<Authorizations>>>
    implements ICanal {
    constructor(props: WithBackContext<CanalComponentProps<Authorizations>>) {
      super(props);
      const { style, ...nextAuthorizations } = props;

      /**
       * @TODO 19-06-01 Find a way to safely forbid the use of reserved prop `style` in a StopName.
       */
      // @ts-ignore

      this.authorizations$.next(nextAuthorizations);
      Navigation.instance.fullScreenDelegate.canalsFullScreenStackProperties$.next(
        this.fullScreenStackProperties$
      );
    }
    static defaultProps = {
      style: StyleSheet.absoluteFill
    };

    canalId = Date.now().toString();

    stopsList = stops.map(stop => ({
      ...stop,
      Component: StopHOC(
        this,
        stop.onBack,
        stop.name,
        stop.Component,
        stop.props
      )
    }));
    authorizations$ = new Subject<Authorizations>();
    progress$ = this.authorizations$.pipe(
      map(authorizations =>
        this.stopsList
          .slice(0)
          .reduce((acc: Stop[], stop, _, stopsCandidatesList) => {
            /**
             * @TODO 19-06-01 Check if Authorization type can be indexed with StopName type.
             * See https://github.com/Microsoft/TypeScript/issues/2491.
             */
            // @ts-ignore
            if (authorizations[stop.name]) {
              acc.push(stop);
            } else {
              stopsCandidatesList.splice(1);
            }
            return acc;
          }, [])
      )
    );

    stack = fromStream(
      this.progress$.pipe(
        map(progressStopsList =>
          progressStopsList.filter(stop => !stop.isFullScreen)
        )
      ),
      []
    );

    fullScreenStackProperties$ = this.progress$.pipe(
      map(progressStopsList => ({
        canalId: this.canalId,
        fullScreenStack: progressStopsList.filter(stop => stop.isFullScreen)
      })),
      distinctUntilChanged(
        (lastFullScreenStackProperties, fullScreenStackProperties) =>
          lastFullScreenStackProperties.fullScreenStack.length ===
          fullScreenStackProperties.fullScreenStack.length
      )
    );

    back$: Observable<IBackEvent> = this.props.backContext.back$.pipe(
      withLatestFrom(this.progress$),
      map(([_, progress]) => {
        const currentStop = last(progress);
        if (currentStop) {
          return { target: currentStop.name };
        }
        return { target: null };
      }),
      share()
    );

    shouldComponentUpdate({
      style,
      ...nextAuthorizations
    }: CanalComponentProps<Authorizations>) {
      /**
       * @TODO 19-06-01 Find a way to safely forbid the use of reserved prop `style` in a StopName.
       */
      // @ts-ignore
      this.authorizations$.next(nextAuthorizations);
      return style !== this.props.style;
    }

    componentWillUnmount() {
      // @ts-ignore @TODO 19-06-01
      this.authorizations$.next({});
      this.authorizations$.complete();
    }

    render() {
      return (
        <View style={this.props.style}>
          <Observer>
            {() =>
              this.stack.current.map(({ name, Component }) => (
                <Component key={name} />
              ))
            }
          </Observer>
        </View>
      );
    }
  }

  // @ts-ignore
  return withBackContext(CanalComponent);
};
