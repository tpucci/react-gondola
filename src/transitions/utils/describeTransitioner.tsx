/* eslint no-undef:0 */

import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Text } from 'react-native';

// @ts-ignore
export function describeTransitioner(Transitioner) {
  // @ts-ignore

  describe(Transitioner.displayName, () => {
    // @ts-ignore

    it('shows its child if direction is forward', () => {
      const testRenderer = TestRenderer.create(
        <Transitioner directionForward>
          <Text testID="child">I should be rendered</Text>
        </Transitioner>
      );
      // @ts-ignore

      expect(testRenderer.toJSON()).toMatchSnapshot();
    });
    // @ts-ignore

    it('hides its child if direction is backward', () => {
      const testRenderer = TestRenderer.create(
        <Transitioner directionForward={false}>
          <Text testID="child">I should not be rendered</Text>
        </Transitioner>
      );
      // @ts-ignore

      expect(testRenderer.toJSON()).toMatchSnapshot();
    });
    // @ts-ignore

    it('snapshots component when direction changed to backward', () => {
      const testRenderer = TestRenderer.create(
        <Transitioner directionForward>
          <Text testID="child">I should not be rendered</Text>
        </Transitioner>
      );
      testRenderer.update(
        <Transitioner directionForward={false}>
          <Text testID="child">I should not be rendered</Text>
        </Transitioner>
      );
      // @ts-ignore

      expect(testRenderer.toJSON()).toMatchSnapshot();
    });
    // @ts-ignore

    it('shows its child if direction changed to forward', () => {
      const testRenderer = TestRenderer.create(
        <Transitioner directionForward={false}>
          <Text testID="child">I should rendered</Text>
        </Transitioner>
      );
      testRenderer.update(
        <Transitioner directionForward>
          <Text testID="child">I should be rendered</Text>
        </Transitioner>
      );
      // @ts-ignore

      expect(testRenderer.toJSON()).toMatchSnapshot();
    });
  });
}
