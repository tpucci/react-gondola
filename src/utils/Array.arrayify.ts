export function arrayify<T>(arg: T | T[]): T[] {
  // @ts-ignore
  if (!arg.length) {
    // @ts-ignore
    return [arg];
  }
  // @ts-ignore
  return arg;
}
