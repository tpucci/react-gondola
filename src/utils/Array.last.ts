export function last<T>(arr: T[]): T | null {
  if (arr.length < 1) { return null; }
  return arr[arr.length - 1];
}
