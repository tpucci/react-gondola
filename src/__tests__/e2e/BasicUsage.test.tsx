import React from 'react';
import TestRenderer from 'react-test-renderer';

import { stopCreator } from '../utils/stopCreator';

import { FullScreenPortal } from '../../FullScreenPortal';
import { createCanal } from '../../createCanal';

describe('BasicUsage', () => {
  it('renders page inside the canal', () => {
    const Canal = createCanal([stopCreator('a'), stopCreator('b')]);
    const testRenderer = TestRenderer.create(
      <FullScreenPortal>
        <Canal a b />
      </FullScreenPortal>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('renders page in the full screen portal', () => {
    const Canal = createCanal([stopCreator('a'), stopCreator('b', true)]);
    const testRenderer = TestRenderer.create(
      <FullScreenPortal>
        <Canal a b />
      </FullScreenPortal>
    );
    testRenderer.update(
      <FullScreenPortal>
        <Canal a b />
      </FullScreenPortal>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
