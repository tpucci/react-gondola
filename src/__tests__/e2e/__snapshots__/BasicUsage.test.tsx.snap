// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicUsage renders page in the full screen portal 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text>
        a
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text>
        b
      </Text>
    </View>
  </View>
</View>
`;

exports[`BasicUsage renders page inside the canal 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text>
        a
      </Text>
    </View>
    <View
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text>
        b
      </Text>
    </View>
  </View>
</View>
`;
